apps/gdalalg_vector_geom_simplify_coverage.cpp

```diff
@@ -0,0 +1,352 @@
+#include "ogr_geos.h"
+#include "ogrsf_frmts.h"
+
+#ifndef _
+#define _(x) (x)
+#endif
+
+//! @cond Doxygen_Suppress
+
+GDALVectorGeomSimplifyCoverageAlgorithm::
+    GDALVectorGeomSimplifyCoverageAlgorithm(bool standaloneStep)
+    : GDALVectorPipelineStepAlgorithm(NAME, DESCRIPTION, HELP_URL,
+                                      standaloneStep)
+{
+    AddActiveLayerArg(&m_activeLayer);
+    AddArg("tolerance", 0, _("Distance tolerance for simplification."),
+           &m_opts.tolerance)
+        .SetRequired()
+        .SetMinValueIncluded(0);
+    AddArg("preserve-boundary", 0,
+           _("Whether the exterior boundary should be preserved."),
+           &m_opts.preserveBoundary);
+}
+
+#if defined HAVE_GEOS &&                                                       \
+    (GEOS_VERSION_MAJOR > 3 ||                                                 \
+     (GEOS_VERSION_MAJOR == 3 && GEOS_VERSION_MINOR >= 12))
+
+/** Class used by vector pipeline steps to create an output on-the-fly
+ * dataset where they can store on-the-fly layers.
+ */
+class GDALVectorGeomSimplifyCoverageOutputDataset final : public GDALDataset
+{
+  public:
+    GDALVectorGeomSimplifyCoverageOutputDataset(GDALDataset &srcDS);
+
+    void AddLayer(std::unique_ptr<OGRLayer> poNewLayer)
+    {
+        m_layers.push_back(std::move(poNewLayer));
+    }
+
+    int GetLayerCount() override
+    {
+        return static_cast<int>(m_layers.size());
+    }
+
+    OGRLayer *GetLayer(int idx) override
+    {
+        return m_layers[idx].get();
+    }
+
+    int TestCapability(const char *) override
+    {
+        // FIXME implement?
+        return false;
+    }
+
+    void ResetReading() override
+    {
+        m_srcDS.ResetReading();
+    }
+
+    OGRFeature *GetNextFeature(OGRLayer **, double *, GDALProgressFunc,
+                               void *) override
+    {
+        // FIXME implement?
+        CPLError(CE_Failure, CPLE_AppDefined,
+                 "GDALDataset::GetNextFeature not implemented.");
+        return nullptr;
+    }
+
+  private:
+    GDALDataset &m_srcDS;
+    std::vector<std::unique_ptr<OGRLayer>> m_layers{};
+};
+
+class GDALVectorGeomSimplifyCoverageAlgorithmLayer final : public OGRLayer
+{
+  public:
+    explicit GDALVectorGeomSimplifyCoverageAlgorithmLayer(
+        OGRLayer &src,
+        const GDALVectorGeomSimplifyCoverageAlgorithm::Options &opts)
+        : m_srcLayer(src), m_opts(opts)
+    {
+    }
+
+    ~GDALVectorGeomSimplifyCoverageAlgorithmLayer() override
+    {
+        if (m_GEOSContext != nullptr)
+        {
+            for (size_t i = 0; i < m_features.size(); i++)
+            {
+                GEOSGeom_destroy_r(m_GEOSContext, m_geos_result[i]);
+            }
+            GEOSFree_r(m_GEOSContext, m_geos_result);
+            finishGEOS_r(m_GEOSContext);
+        }
+    }
+
+    int TestCapability(const char *pszCap) override
+    {
+        if (EQUAL(pszCap, OLCRandomRead))
+        {
+            return true;
+        }
```

Comment:
```suggestion
```

Reply:
If you keep OLCRandomRead, then you should implement GetFeature(), which you can actually implement by materializing and then returning m_features[idx]